src/freenas/usr/local/bin/truenas-autotune.py

```diff
@@ -0,0 +1,159 @@
+#!/usr/bin/env python3
+import sys
+from types import SimpleNamespace
+
+import psutil
+
+from middlewared.plugins.system.dmi import DMIDecode
+from middlewared.plugins.system.product import is_enterprise_ix_hardware
+from middlewared.plugins.truenas import get_chassis_hardware
+from middlewared.plugins.tunables import zfs_parameter_value
+from middlewared.utils.db import query_table, update_table
+
+KiB = 1024 ** 1
+MiB = 1024 ** 2
+GiB = 1024 ** 3
+
+MIN_ZFS_RESERVED_MEM = 1 * GiB
+
+zfs_parameters = {}
+
+
+def zfs_parameter(tunable_name):
+    def decorator(func):
+        zfs_parameters[tunable_name] = func
+        return func
+
+    return decorator
+
+
+@zfs_parameter("zfs_dirty_data_max_max")
+def guess_vfs_zfs_dirty_data_max_max(context):
+    if context.hardware.startswith("M"):
+        return 12 * GiB
+    else:
+        return None
+
+
+@zfs_parameter("zfs_arc_max")
+def guess_vfs_zfs_arc_max(context):
+    if context.physmem / GiB > 200:
+        return int(max(min(int(context.physmem * .92),
+                           context.physmem - (context.kernel_reserved + context.userland_reserved)),
+                       MIN_ZFS_RESERVED_MEM))
+    else:
+        return int(max(min(int(context.physmem * .9),
+                           context.physmem - (context.kernel_reserved + context.userland_reserved)),
+                       MIN_ZFS_RESERVED_MEM))
+
+
+@zfs_parameter("l2arc_noprefetch")
+def guess_vfs_zfs_l2arc_noprefetch(context):
+    return 0
+
+
+@zfs_parameter("l2arc_write_max")
+def guess_vfs_zfs_l2arc_write_max(context):
+    return 10000000
+
+
+@zfs_parameter("l2arc_write_boost")
+def guess_vfs_zfs_l2arc_write_boost(context):
+    return 40000000
+
+
+@zfs_parameter("zfs_vdev_async_read_max_active")
+def guess_vfs_zfs_vdev_async_read_max_active(context):
+    if context.hardware == "Z50":
```

Comment:
`context.hardware` will never be `Z50` only. It will be `TRUENAS-{Z50/M30/X20}` etc and sometimes we even tagged SMBIOS ending with `-HA` (i.e. `TRUENAS-M50-HA`). The check needs to be `context.hardware.startswith('TRUENAS-Z50')`.